Jenkinsfile

```diff
@@ -0,0 +1,291 @@
+                throw err
+            } finally {
+                // clean up
+                step([$class: 'WsCleanup'])
+            }
+        }
+    }
+}
+
+
+def baseBuild() {
+    dir('mbed-flasher') {
+        // checkout scm
+        echo "checkout scm start"
+
+        def scmVars = checkout scm
+        env.GIT_COMMIT_HASH = scmVars.GIT_COMMIT
+
+
+        // copy artifacts
+        copyArtifacts()
+
+        if (isUnix()) {
+            stage("linux py2") {
+                unittest("py2")
+            }
+
+            stage("linux py3") {
+                unittest("py3")
+            }
+
+            stage("pylint check") {
+                linux_pylint_check()
+            }
+
+            postBuild()
+        }
+        else {
+            stage("windows py2") {
+                winTest("py2")
+            }
+
+            stage("windows py3") {
+                winTest("py3")
+            }
+        }
+    }
+}
+
+
+def copyArtifacts(){
+    echo "Copying artifacts from mbed-os-cliapp_for_CI/label=GCC_ARM"
+    step([
+        $class: 'CopyArtifact',
+        fingerprintArtifacts: true,
+        flatten: true,
+        projectName: "ARMmbed/mbed-os-cliapp/master",
+        filter: '**/K64F_arm-none-eabi-gcc_eth0/mbed-os-cliapp.bin',
+        selector: [
+            $class: 'StatusBuildSelector',
+            stable: true
+        ],
+        target: 'test/'
+    ])
+}
+
+
+def setBuildStatus(String state, String context, String message) {
+    step([
+        $class: "GitHubCommitStatusSetter",
+        reposSource: [
+            $class: "ManuallyEnteredRepositorySource",
+            url: "https://github.com/ARMmbed/mbed-flasher.git"
+        ],
+        contextSource: [
+            $class: "ManuallyEnteredCommitContextSource",
+            context: context
+        ],
+        errorHandlers: [[
+            $class: "ChangingBuildStatusErrorHandler",
+            result: "UNSTABLE"
+        ]],
+        commitShaSource: [
+            $class: "ManuallyEnteredShaSource",
+            sha: env.GIT_COMMIT_HASH
+        ],
+        statusResultSource: [
+            $class: 'ConditionalStatusResultSource',
+            results: [
+                [
+                    $class: 'AnyBuildResult',
+                    message: message,
+                    state: state
+                ]
+            ]
+        ]
+    ])
+}
+
+
+def unittest(String pythonVersion) {
+    echo "set ${pythonVersion} github status"
+    String buildName = "linux ${pythonVersion} test"
+    setBuildStatus('PENDING', "${buildName}", 'test start')
+    try {
+        if (pythonVersion == "py3") {
+            // create python 3 venv
+            sh """
+                python3 -m venv .py3venv --without-pip
+                . .py3venv/bin/activate
+                curl https://bootstrap.pypa.io/get-pip.py | python
+                pip install coverage mock astroid==1.5.3 pylint==1.7.2
+                pip install -I mbed-ls==1.2.14
```

Comment:
should not do this

Reply:
we have to now, if we specify mbed-ls version in setup.py, it never know which is correct version.

for mbed-ls relase, I guess it doesn't consider much about compatibility with windows installation and python3.

1.2.16 and 1.2.15 cannot doesn't support python3
1.2.14  cannot install on windows successfully

Reply:
otherwise, we  should disable python3 for linux and windows.

Reply:
This is an mbedls bug then. We can not alter the test environment to get positive result.

Reply:
Should probably specify 1.2.14 and create a bug for mbedls.

Reply:
Or if you want to use different versions based on platform, it should be done in setup.py.

Reply:
yep, there is an issue raised for mbed-ls [here](https://github.com/ARMmbed/mbed-ls/issues/226).
okay, I did it in setup.py